src/Codegen/Codegen.hs

```diff
@@ -470,34 +480,53 @@ genExpr a@(I.App _ _ ty) = do
               ++ [[cexp|&$exp:tmp|]]
           call = [citems|$exp:(activate $ fnEnter `ccall` enterArgs);|]
       return (tmp, concat evalStms ++ call ++ yield)
-    (I.Data tg dty) -> do
-      info <- gets fnAdtInfo
-      case M.lookup tg (isPointer info) of
-        Nothing    -> fail "Couldn't find tag"
-        Just False -> fail "Cannot handle integer types with fields yet"
-        Just True  -> do
-          tmp <- genTmp dty
-          let Just sz = M.lookup tg (dconSize info)
-          let alloc   = [[citem|$exp:tmp = $exp:(new_adt sz tg);|]]
-          let initField y i = [citem|$exp:(adt_field tmp i) = $exp:y;|]
-              initFields = zipWith initField argVals [0 ..]
-          return (tmp, concat evalStms ++ alloc ++ initFields)
+    (I.Data dcon dty) -> do
+      onHeap <- getsDCon dconOnHeap dcon
+      unless onHeap $ do
+        fail $ "Cannot handle packed fields yet, for: " ++ show dcon
+      construct <- getsDCon dconConstruct dcon
+      destruct  <- getsDCon dconDestruct dcon
+      tmp       <- genTmp dty
+      let alloc = [[citem|$exp:tmp = $exp:construct;|]]
+          initField y i = [citem|$exp:(destruct i tmp) = $exp:y;|]
+          initFields = zipWith initField argVals [0 ..]
+      return (tmp, concat evalStms ++ alloc ++ initFields)
     _ -> fail $ "Cannot apply this expression: " ++ show fn
 genExpr (I.Match s as t) = do
-  -- TODO: document this
+  -- To implement a match expression, we need to generate a C switch statement
+  --
+  -- which switches on the tag of the scrutinee @s@.
+  -- However, since we're already using a switch statement for jumping to
+  -- program counters, we cannot simply nest the switch statement, since we will
+  -- risk the inner switch statement (for the match expression) shadowing the
+  -- cases of the outer switch statement (for the program counters). In
+  -- particular, we will be unable to yield inside of match expression arm with
+  -- this naive compilation scheme.
+  --
+  -- So, in order to keep the C statements "flat", we use a basic block-style
+  -- scheme, where the inner switch statement only jumps (using @goto@) to
+  -- blocks corresponding to each arm of the match expression; this ensures that
+  -- we never need to yield in the inner switch statement, thus side-stepping
+  -- C's limitation. At the end of each block, we jump to a join statement that
+  -- follows the blocks generated for each arm.
   (sExp, sStms) <- genExpr s
   scrut         <- genTmp $ extract s
   val           <- genTmp t
   joinLabel     <- freshLabel
 
   let
     assignScrut = [citems|$exp:scrut = $exp:sExp;|]
-    tag         = adt_tag scrut
+    tag         = adt_tag scrut -- TODO: look this up using typeScrut
```

Comment:
Is this TODO here for later when packed value ADTs need unique tag lookup functions?

Reply:
Yes! That's exactly right.